keysync/server.go

```diff
@@ -206,9 +228,30 @@ func (ks *KeySyncServer) syncSecretsToLocalKeys(secList *corev1.SecretList, skh
 			path := filepath.Join(ks.keySyncDir, filename)
 			if !fileExists(path) {
 				logrus.Printf("Syncing new key: %v", filename)
-				err := ioutil.WriteFile(path, data, 0600)
+				err := ioutil.WriteFile(path, data, ks.keyFilePermissions)
+				if err != nil {
+					logrus.Errorf("Unable to write file %s: %v", path, err)
+					continue
+				}
+			}
+
+			// Permission and ownership corrigation if needed
+			fileInfo, err := os.Stat(path)
+			if err != nil {
+				logrus.Errorf("Unable to get file %s: %v", path, err)
+			}
+			if fileInfo.Mode() != ks.keyFilePermissions {
+				err = os.Chmod(path, ks.keyFilePermissions)
+				if err != nil {
+					logrus.Errorf("Unable to change file permissions %s: %v", path, err)
+					continue
+				}
+			}
+			if ((ks.keyFileOwnerUID != -1) && (fileInfo.Sys().(*syscall.Stat_t).Uid != uint32(ks.keyFileOwnerUID))) ||
+				((ks.keyFileOwnerGID != -1) && (fileInfo.Sys().(*syscall.Stat_t).Gid != uint32(ks.keyFileOwnerGID))) {
+				err = os.Chown(path, ks.keyFileOwnerUID, ks.keyFileOwnerGID)
```

Comment:
I'm finding it a bit difficult to figure out the usecase of setting a UID/GID instead of using the one from the application. I feel like this introduces a lot of privileges. I would prefer not to need to require `CAP_CHOWN`. These are created within the UTS namespace of the container, and UID/GID should be fairly transparent. Is there a specific usecase im missing? 

OpenShift does UID/GID mapping so this may become a bit weird.

Reply:
Fair question, our use case is a bit tricky.

We want the private keys persisted on the nodes' file system to be owned by `root:root` and not some random UID:GID, as the content is sensitive (and they keys will became essential components of the cluster operation).

In order to write the files as `root:root` we have two options:
- run the container as privileged (we don't want that for sure!),
- give `CAP_CHOWN` to the process, so it can change the permissions of these files.

> I feel like this introduces a lot of privileges.

I agree, on one side we win something, on the other we lose.

> OpenShift does UID/GID mapping so this may become a bit weird.

If the parameters are not specified, the files will be created with the process UID:GID.
(In our case we are going to use an SCC that allows specifying any UID/GID and will not rely on OpenShift mapping.)